src/main/java/graphql/ProfilerImpl.java

```diff
@@ -0,0 +1,148 @@
+package graphql;
+
+import graphql.execution.EngineRunningObserver;
+import graphql.execution.ExecutionId;
+import graphql.execution.ResultPath;
+import graphql.execution.instrumentation.ChainedInstrumentation;
+import graphql.execution.instrumentation.Instrumentation;
+import graphql.execution.instrumentation.dataloader.DataLoaderDispatchingContextKeys;
+import graphql.language.OperationDefinition;
+import graphql.schema.DataFetcher;
+import graphql.schema.PropertyDataFetcher;
+import graphql.schema.SingletonPropertyDataFetcher;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicLong;
+
```

Comment:
Not for this PR, but in the future, with more Profiler features you might want to introduce options to toggle sets of tracking on and off

e.g. toggle DataLoader related tracking on and off